From: Miao Xie <miaox@cn.fujitsu.com>
Date: Wed, 13 Apr 2011 14:07:59 +0800
Patch-mainline: yes
References: FATE#306586
Subject: [PATCH] Btrfs: Check validity before setting an acl

Call posix_acl_valid() to check if an acl is valid or not.

Signed-off-by: Miao Xie <miaox@cn.fujitsu.com>
Signed-off-by: Li Zefan <lizf@cn.fujitsu.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/acl.c |    9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

--- a/fs/btrfs/acl.c
+++ b/fs/btrfs/acl.c
@@ -177,16 +177,17 @@ static int btrfs_xattr_set_acl(struct in
 
 	if (value) {
 		acl = posix_acl_from_xattr(value, size);
-		if (acl == NULL) {
-			value = NULL;
-			size = 0;
+		if (acl) {
+			ret = posix_acl_valid(acl);
+			if (ret)
+				goto out;
 		} else if (IS_ERR(acl)) {
 			return PTR_ERR(acl);
 		}
 	}
 
 	ret = btrfs_set_acl(NULL, inode, acl, type);
-
+out:
 	posix_acl_release(acl);
 
 	return ret;
